### DISCUSSION TITLE
How to duplicate a collection of SDAs across a set of resources and visualize all asset lineages?

### DISCUSSION THREAD

**Discussion by rprechelt** [dagster-io/dagster#18353] [Q&A] [answered]
What is the recommended approach for duplicating & materializing a graph-backed set of assets across multiple implementations of a resource in one Dagster instance? 

(FYI: new to Dagster) I have a set of SDAs backed by a graph (that uses a `MultiPartition`) that takes a resource argument in the top-level `Definitions`. I want to generate and track these SDAs across N~3 different implementations of this resource and (ideally) have them all be simultaneously be visible in the Asset Lineage as N=3 copies of the graph/SDAs. In effect, I want to effectively create a "third dimension" to the `MultiPartition` by duplicating over different resource implementations (since `MultiPartition` is limited to 2D)

### Context
We are evaluating whether we can port our existing task-based DAG pipeline to Dagster SDAs. We currently use a non-Dagster DAG that takes `(date x name x source)` as _input,_ with roughly ~50M unique input combinations (3000 x 6000 x 3) that produce roughly ~1B output files (although it is rarely rerun from scratch). The DAG that produces the assets are identical for each input tuple where `source` really acts like a Resource and determines where to pull the original data from. In our current Dagster experiment, we mapped the `(date x name) `dimensions down to Dagster SDAs with a `MultiPartitionDefinition[DailyPartitionDef, StaticPartitionDef]` with roughly ~18M unique partition keys (since `MultiPartition`'s are limited to 2D). Is there a better way to map this existing DAG structure down to Dagster? 

### What we've considered
+ I understand `RunConfigurations` can be used to adjust resources depending upon the environment (i.e. Dev vs Prod) but as far as I can see, there is still only one "set of assets" and they aren't stored simultaneously for both configurations on the same Dagster instance (i.e. they aren't duplicated over `RunConfigurations`—correct me if I'm wrong).
+ Dynamic Graphs also don't seem to allow what we need—as far as I can see, they just allow for generating an undetermined number of outputs that are then combined later on in the Dagster graph and not for duplicating graphs.

I don't need working code samples but if someone could point me in the right direction for how best to achieve this behavior (or similar), that would be great! Any general recommendations would also be appreciated since I am new to Dagster.

**Comment by smackesey** [comment, marked as answer]
Hi @rprechelt, thanks for the clear problem statement. I think the ideal solution would be a three-dimensional partition but as you've seen, we don't support that yet.

I recommend that you create a separate copy of the asset definitions for each resource and distinguish them using a different key prefix and asset group on each. Something like this:

```
def my_asset_impl():
    ...

@asset(group_name="foo", key=AssetKey(["foo", "my_asset"]))
def foo_my_asset(context: AssetExecutionContext):
    my_asset_impl()

@asset(group_name="bar", key=AssetKey(["bar", "my_asset"]))
def bar_my_asset(context: AssetExecutionContext):
    my_asset_impl()

defs = Definitions(
    assets=[foo_my_asset, bar_my_asset],
)
```
